src/Polly.Core/CircuitBreaker/Controller/CircuitStateController.cs

```diff
@@ -0,0 +1,343 @@
+    {
+        _breakDuration = options.BreakDuration;
+        _onOpened = options.OnOpened.CreateHandler();
+        _onClosed = options.OnClosed.CreateHandler();
+        _onHalfOpen = options.OnHalfOpened.CreateHandler();
+        _behavior = behavior;
+        _timeProvider = timeProvider;
+        _telemetry = telemetry;
+    }
+
+    public CircuitState CircuitState
+    {
+        get
+        {
+            EnsureNotDisposed();
+
+            lock (_lock)
+            {
+                return _circuitState;
+            }
+        }
+    }
+
+    public Exception? LastException
+    {
+        get
+        {
+            EnsureNotDisposed();
+
+            lock (_lock)
+            {
+                return _lastOutcome?.Exception;
+            }
+        }
+    }
+
+    public Outcome? LastHandledOutcome
+    {
+        get
+        {
+            EnsureNotDisposed();
+
+            lock (_lock)
+            {
+                return _lastOutcome;
+            }
+        }
+    }
+
+    public ValueTask IsolateCircuitAsync(ResilienceContext context)
+    {
+        EnsureNotDisposed();
+
+        context.Initialize<VoidResult>(isSynchronous: false);
+
+        Task? task;
+
+        lock (_lock)
+        {
+            SetLastHandledOutcome_NeedsLock(new Outcome<VoidResult>(new IsolatedCircuitException()));
+            OpenCircuitFor_NeedsLock(new Outcome<VoidResult>(VoidResult.Instance), TimeSpan.MaxValue, manual: true, context, out task);
+            _circuitState = CircuitState.Isolated;
+        }
+
+        return ExecuteScheduledTaskAsync(task, context);
+    }
+
+    public ValueTask CloseCircuitAsync(ResilienceContext context)
+    {
+        EnsureNotDisposed();
+
+        context.Initialize<VoidResult>(isSynchronous: false);
+
+        Task? task;
+
+        lock (_lock)
+        {
+            CloseCircuit_NeedsLock(new Outcome<VoidResult>(VoidResult.Instance), manual: true, context, out task);
+        }
+
+        return ExecuteScheduledTaskAsync(task, context);
+    }
+
+    public async ValueTask OnActionPreExecuteAsync(ResilienceContext context)
+    {
+        EnsureNotDisposed();
+
+        Exception? exception = null;
+        bool isHalfOpen = false;
+
+        Task? task = null;
+
+        lock (_lock)
+        {
+            // check if circuit can be half-opened
+            if (_circuitState == CircuitState.Open && PermitHalfOpenCircuitTest_NeedsLock())
+            {
+                _circuitState = CircuitState.HalfOpen;
+                _telemetry.Report(CircuitBreakerConstants.OnHalfOpenEvent, new OnCircuitHalfOpenedArguments(context));
+                isHalfOpen = true;
+            }
+
+            exception = _circuitState switch
+            {
+                CircuitState.Open => GetBreakingException_NeedsLock(),
+                CircuitState.HalfOpen when isHalfOpen is false => GetBreakingException_NeedsLock(),
+                CircuitState.Isolated => new IsolatedCircuitException(),
```

Comment:
I wonder if we could add some context to this exception so people could more easily determine which one was isolated, rather than relying on the stack trace to try and work it out.

Reply:
What context do you have in mind? 

Storing `ResilienceContext` on exception is problematic as it will be returned to pool.

Reply:
At lease for us we haven't really needed this info. We can add it later without breaking changes if there comes a feature request.